### PR TITLE
xds: use typedstruct

### PR DESCRIPTION
**Pull request by kyessenov**
Signed-off-by: Kuat Yessenov <kuat@google.com>

Struct is deprecated, use UDPA struct.

### CODE REVIEW

**Comment by bianpengyuan** [comment]
Can we also update the filter in installer? https://github.com/istio/installer/blob/2d8249df15868ab1bdd116d5368285613a4fd623/istio-control/istio-discovery/templates/telemetryv2_1.5.yaml

File: tests/integration/telemetry/stats/prometheus/testdata/metadata_exchange_filter.yaml [hunk 1]
```diff
@@ -15,11 +15,14 @@ spec:
...
           name: envoy.filters.http.wasm
-          config:
-            config:
-              configuration: envoy.wasm.metadata_exchange
-              vm_config:
-                runtime: envoy.wasm.runtime.null
-                code:
-                  local:
-                    inline_string: envoy.wasm.metadata_exchange
+          typed_config:
```
Comment (bianpengyuan) [inline tests/integration/telemetry/stats/prometheus/testdata/metadata_exchange_filter.yaml]: should this be
```
config:
  typed_config:
     ...
```

error log:
```
Exiting due to setup failure: exit status 1: Error from server: error when creating "/logs/artifacts/stats-filter-test-a94367dbd0d14/_suite_context/galley-workdir-713754910/networking_istio_io_EnvoyFilter_istio-bookinfo-1-32783_stats-filter-3.yaml": admission webhook "pilot.validation.istio.io" denied the request: configuration is invalid: 3 errors occurred:
	* Envoy filter: unknown message type "udpa.type.v1.TypedStruct"
	* Envoy filter: unknown message type "udpa.type.v1.TypedStruct"
	* Envoy filter: unknown message type "udpa.type.v1.TypedStruct"
```

Comment (kyessenov) [reply tests/integration/telemetry/stats/prometheus/testdata/metadata_exchange_filter.yaml]: no, it should be typed_config at the top level. the error is due to galley not linking some UDPA protos. trying to move static linking into shared code.

**Comment by kyessenov** [comment]
@bianpengyuan Yes, I'll update installer repo too. Checking that it's all working first here.

File: tests/integration/telemetry/stats/prometheus/testdata/metadata_exchange_filter.yaml (hunk 1 above)
Comment (kyessenov) [reply tests/integration/telemetry/stats/prometheus/testdata/metadata_exchange_filter.yaml]: OMFG. I think this is again gogoproto vs goproto madness. I'll see if I can workaround it or not.

**Comment by mandarjog** [comment]
@kyessenov does anything still depend on gogoproto?
If not now is the time to get rid of it.

**Comment by kyessenov** [comment]
@mandarjog Of course, anything to do with galley is entirely dependent on gogo. Last time I checked with Oz and Martin, that was a massive effort to drop gogo.

**Comment by kyessenov** [comment]
Tests may pass and hopefully that means we're OK with this. But I admit I don't understand why that works. The code for envoyfilter just does dynamic casts from gogo to goproto which is not kosher.

**Comment by howardjohn** [comment]
I thought we were on gogo for something related to mixer performance? It seem like switching galley from gogo to golang shouldn't be too hard other than a painful PR to rename everything like when we switched xds?

**Comment by mandarjog** [comment]
There is/was gogo for datapath in mixer , which should be orthogonal to anything control plane. But it looks like galley is using some gogoisms .

**Comment by kyessenov** [comment]
@howardjohn It could be, but I'm not an expert in that area. I pitched the idea to Oz/Martin and they wanted me to not drop gogo for galley so I didn't pursue it further.

**Comment by mandarjog** [comment]
Gogo is found 300 times in our codebase . Hard to gauge gogo specific code. @ozevren @geeknoid can you shed some light on which parts need gogo and why. Thnx.   @ayj

**Comment by ozevren** [comment]
The biggest problem with gogo v.s. golang & Galley was the trying to support dual mode, where some protos were in gogo, and some in golang. That permeated everywhere. Galley is now in gogo-only mode, and most code assumes it depends on gogo libraries. However, it shouldn't be using gogo specific features. 

It is conceptually easy to port Galley to be golang proto only. However, it is probably quite tedious. Tests should give pretty good coverage. Still, it is probably a good idea to make sure point-wise things work properly (e.g. struct based configuration, handling of anys and datetimes etc.).

**Pull request merged**